src/java.base/share/classes/java/io/Reader.java

```diff
@@ -140,6 +140,124 @@ public void close() {
+     * <p> After the reader has been closed, the {@code read()},
+     * {@code read(char[])}, {@code read(char[], int, int)},
+     * {@code read(CharBuffer)}, {@code ready()}, {@code skip(long)}, and
+     * {@code transferTo()} methods all throw {@code IOException}.
+     *
+     * <p> The {@code markSupported()} method returns {@code true}.
+     *
+     * @param source {@code CharSequence} providing the character stream.
+     *
+     * @return a {@code Reader} which reads characters from {@code source}
+     *
+     * @throws NullPointerException if {@code source} is {@code null}
+     *
+     * @since 24
+     */
+    public static Reader of(CharSequence source) {
+        Objects.requireNonNull(source);
+
+        return new Reader() {
+            private final int length = source.length();
+            private CharSequence cs = source;
+            private int next = 0;
+            private int mark = 0;
+
+            /** Check to make sure that the stream has not been closed */
+            private void ensureOpen() throws IOException {
+                if (cs == null)
+                    throw new IOException("Stream closed");
+            }
+
+            @Override
+            public int read() throws IOException {
+                ensureOpen();
+                if (next >= length)
+                    return -1;
+                return cs.charAt(next++);
+            }
+
+            @Override
+            public int read(char[] cbuf, int off, int len) throws IOException {
+                ensureOpen();
+                Objects.checkFromIndexSize(off, len, cbuf.length);
+                if (len == 0) {
+                    return 0;
+                }
+                if (next >= length)
+                    return -1;
+                int n = Math.min(length - next, len);
+                switch (cs) {
+                    case String s -> s.getChars(next, next + n, cbuf, off);
+                    case StringBuilder sb -> sb.getChars(next, next + n, cbuf, off);
+                    case StringBuffer sb -> sb.getChars(next, next + n, cbuf, off);
+                    case CharBuffer cb -> cb.get(next, cbuf, off, n);
+                    default -> {
+                        for (int i = 0; i < n; i++)
+                            cbuf[off + i] = cs.charAt(next + i);
+                    }
+                }
+                next += n;
+                return n;
+            }
+
+            @Override
+            public long skip(long n) throws IOException {
+                ensureOpen();
+                if (next >= length)
+                    return 0;
+                // Bound skip by beginning and end of the source
+                long r = Math.min(length - next, n);
+                r = Math.max(-next, r);
+                next += (int)r;
+                return r;
+            }
+
+            @Override
+            public boolean ready() throws IOException {
+                ensureOpen();
+                return true;
```

Comment:
I guess it’s not wrong, but can this be false when end of source is reached? Related to this you already described it is not synchronized but should we also describe it does not support changes to the length during “iteration” (could be in the same thread).

Reply:
Ignore my “thinko”/comment. of course it needs to return true for the EOD.